src/test/java/htsjdk/samtools/BAMFileWriterTest.java

```diff
@@ -229,4 +243,179 @@ public void testBAMReadBases() throws IOException {
+        builder.addPair("pair1", 0, 1, 100_000, false, false, cigar.toString(), cigar.toString(), true, false, 30);
+
+        testHelper(builder, SAMFileHeader.SortOrder.coordinate, true);
+    }
+
+    @DataProvider
+    public Object[][] sentinelCigarData() {
+        return new Object[][]{
+                {1, 0, 1},
+                {0, 1, 1},
+                {0, 0, 0},
+        };
+    }
+
+    @Test(dataProvider = "sentinelCigarData", expectedExceptions = IllegalStateException.class)
+    public void testWrongCGTagCigars(final int readOffset, final int refOffset, final int cigarOpsOffset) throws Exception {
+        final SAMRecordSetBuilder builder = new SAMRecordSetBuilder(true, SAMFileHeader.SortOrder.coordinate);
+        final List<CigarOperator> operators = getCigarOperatorsForTest(BAMRecord.MAX_CIGAR_OPERATORS + cigarOpsOffset);
+        final Cigar cigar = Cigar.fromCigarOperators(operators);
+        final int[] cigarEncoding = BinaryCigarCodec.encode(cigar);
+
+        final Cigar sentinelCigar = BAMRecordCodec.makeSentinelCigar(cigar);
+        final List<CigarElement> sentinelCigarElements = new ArrayList<>(2);
+
+        sentinelCigarElements.add(new CigarElement(
+                sentinelCigar.getCigarElement(0).getLength() + readOffset,
+                sentinelCigar.getCigarElement(0).getOperator()));
+        sentinelCigarElements.add(new CigarElement(
+                sentinelCigar.getCigarElement(1).getLength() + refOffset,
+                sentinelCigar.getCigarElement(1).getOperator()));
+
+        final String sentinelCigarString = new Cigar(sentinelCigarElements).toString();
+
+        builder.addPair("pair1", 0, 1, 100_000, false, false, sentinelCigarString, sentinelCigarString, true, false, 30);
+
+        final SAMRecord frag = builder.addFrag("frag1", 0, 1, false, false, sentinelCigarString, null, 30);
+        frag.setAttribute(SAMTag.CG.name(), cigarEncoding);
+
+        final List<SAMRecord> pairOfReads = builder.addPair("pair1", 0, 1, 100_000, false, false, cigar.toString(), cigar.toString(), true, false, 30);
+        for (final SAMRecord rec : pairOfReads) {
+            rec.setAttribute(SAMTag.CG.name(), cigarEncoding);
+        }
+
+        final File bamFile = File.createTempFile("test.", BamFileIoUtils.BAM_FILE_EXTENSION);
+        bamFile.deleteOnExit();
+
+        try (final SAMFileWriter bamWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(builder.getHeader(), false, bamFile)) {
+            for (final SAMRecord record : builder.getRecords())
+                bamWriter.addAlignment(record);
+        }
+
+        try (final SamReader reader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT).open(bamFile)) {
+            reader.iterator().forEachRemaining(SAMRecord::getCigar);
+        }
+    }
+
+    @Test(dataProvider = "longCigarsData")
+    public void testNoCGTagCigars(final int numOps) throws Exception {
```

Comment:
I think we should add tests to show that we don't have an incorrectly empty cigar if clearAttributes is called, and also test that we don't end up with an CG tag in the case where setCigar or setCigarString is called.

Reply:
lets discuss these cases in person.